File: Lab3IvaniukKI304/src/Lab3IvaniukKI304/Pond.java
package Lab3IvaniukKI304;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;

/**
 * The {@code Pond} class represents a pond with various attributes and behaviors.
 * It can log the pond's activities to a text file.
 * This class is abstract and should be extended by specific water bodies like Sea.
 *
 * @author Ivaniuk Dmytro
 */
public abstract class Pond {
    protected int length;
    protected int width;
    protected int depth;
    protected String logFileName;
    protected FileWriter logFileWriter;

    /**
     * Constructs a new pond with default values and initializes a log file.
     */
    public Pond() {
        this.length = 0;
        this.width = 0;
        this.depth = 0;
        logFileName = "pondLog.txt";
        try {
            logFileWriter = new FileWriter(new File(logFileName));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    /**
     * Constructs a new pond with the specified length, width, and depth, and initializes a log file.
     *
     * @param length The length of the pond.
     * @param width  The width of the pond.
     * @param depth  The depth of the pond.
     */
    public Pond(int length, int width, int depth) {
        this.length = length;
        this.width = width;
        this.depth = depth;
        logFileName = "pondLog.txt";
        try {
            logFileWriter = new FileWriter(new File(logFileName));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void setLength(int length) {
        this.length = length;
    }

    public void setWidth(int width) {
        this.width = width;
    }

    public void setDepth(int depth) {
        this.depth = depth;
    }

    public int getLength() {
        return length;
    }

    public int getWidth() {
        return width;
    }

    public int getDepth() {
        return depth;
    }

    public void area() {
        log("Area: " + length * width + " km^2");
    }

    public void volume() {
        log("Volume: " + length * width * depth + " km^3");
    }

    public void perimeter() {
        log("Perimeter: " + 2 * (length + width) + " km");
    }

    public void maxValue() {
        log("Max value: " + Math.max(length, Math.max(width, depth)));
    }

    public void scale(int scaleFactor) {
        this.length *= scaleFactor;
        this.width *= scaleFactor;
        this.depth *= scaleFactor;
        log("Scales each parameter by " + scaleFactor);
    }

    public void calculateVolumeByPercentage(double percentage) {
        if (percentage < 0 || percentage > 100) {
            System.out.println("The percentage must be between 0 and 100.");
        } else {
            log("Volume(" + percentage + "%): " + length * width * depth * (percentage / 100) + " km^3");
        }
    }

    public void determineDepth() {
        if (depth >= 1) {
            log("The pond is deep. Its depth is " + depth + " km.");
        } else {
            log("The pond is shallow. Its depth is " + depth + " km.");
        }
    }

    public void log(String message) {
        try {
            logFileWriter.write(message + "\n");
            logFileWriter.flush();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void dispose() throws IOException {
        logFileWriter.close();
    }

    // Abstract method that needs to be implemented by subclasses.
    public abstract void uniqueFeature();
}

File: Lab9IvaniukKI304/pond.py
import os

class Pond:
    def __init__(self, length=0, width=0, depth=0):
        """
        Базовий клас для водойм.
        :param length: Довжина водойми
        :param width: Ширина водойми
        :param depth: Глибина водойми
        """
        self.length = length
        self.width = width
        self.depth = depth
        self.log_file_name = "pond_log.txt"
        self._log_file = open(self.log_file_name, "w")

    def set_length(self, length):
        self.length = length

    def set_width(self, width):
        self.width = width

    def set_depth(self, depth):
        self.depth = depth

    def get_length(self):
        return self.length

    def get_width(self):
        return self.width

    def get_depth(self):
        return self.depth

    def area(self):
        result = self.length * self.width
        self._log(f"Area: {result} km^2")
        return result

    def volume(self):
        result = self.length * self.width * self.depth
        self._log(f"Volume: {result} km^3")
        return result

    def perimeter(self):
        result = 2 * (self.length + self.width)
        self._log(f"Perimeter: {result} km")
        return result

    def max_value(self):
        result = max(self.length, self.width, self.depth)
        self._log(f"Max value: {result}")
        return result

    def scale(self, scale_factor):
        self.length *= scale_factor
        self.width *= scale_factor
        self.depth *= scale_factor
        self._log(f"Scaled dimensions by {scale_factor}")

    def calculate_volume_by_percentage(self, percentage):
        if not (0 <= percentage <= 100):
            print("The percentage must be between 0 and 100.")
            return None
        result = self.length * self.width * self.depth * (percentage / 100)
        self._log(f"Volume ({percentage}%): {result} km^3")
        return result

    def determine_depth(self):
        if self.depth >= 1:
            self._log(f"The pond is deep. Its depth is {self.depth} km.")
        else:
            self._log(f"The pond is shallow. Its depth is {self.depth} km.")

    def _log(self, message):
        self._log_file.write(message + "\n")
        self._log_file.flush()

    def dispose(self):
        self._log_file.close()
        if os.path.exists(self.log_file_name):
            print(f"Log saved to {self.log_file_name}")

    def unique_feature(self):
        """Абстрактний метод, який реалізується у похідних класах."""
        raise NotImplementedError("This method should be implemented by subclasses.")
